src/components/mask-input/mask-input.ts

```diff
@@ -0,0 +1,380 @@
+    if (!this.hasFocus && !this._value) {
+      this.maskedValue = this.parser.apply();
+    }
+  }
+
+  protected handleDragLeave() {
+    if (!this.hasFocus) {
+      this.updateMaskedValue();
+    }
+  }
+
+  protected override handleFocus() {
+    this.hasFocus = true;
+    super.handleFocus();
+
+    if (this.readonly) {
+      return;
+    }
+
+    if (!this._value) {
+      this.maskedValue = this.parser.apply();
+      // In case of empty value, select the whole mask
+      this.updateComplete.then(() => this.select());
+    }
+  }
+
+  protected override handleBlur() {
+    this.hasFocus = false;
+    this.updateMaskedValue();
+    super.handleBlur();
+  }
+
+  protected override handleChange() {
+    this.emitEvent('igcChange', { detail: this.value });
+    this.invalid = !this.checkValidity();
+  }
+
+  protected handleInvalid() {
+    this.invalid = true;
+  }
+
+  protected handleClick() {
+    // Clicking at the end of the input field will select the entire mask
+    if (
+      this.input.selectionStart === this.input.selectionEnd &&
+      this.input.selectionStart === this.maskedValue.length
+    ) {
+      this.select();
+    }
+  }
+
+  protected updateMaskedValue() {
+    if (this.maskedValue === this.parser.apply()) {
+      this.maskedValue = '';
+    }
+  }
+
+  /** Replaces the selected text in the control and re-applies the mask */
+  public override setRangeText(
+    replacement: string,
+    start: number,
+    end: number,
+    _selectMode: 'select' | 'start' | 'end' | 'preserve' = 'preserve'
+  ) {
+    this.input.value = this.parser.replace(
+      this.input.value,
+      replacement,
+      start,
+      end
+    ).value;
+    this.maskedValue = this.parser.apply(this.parser.parse(this.input.value));
+    this._value = this.parser.parse(this.maskedValue);
+  }
+
+  public override setSelectionRange(
+    start: number,
+    end: number,
+    direction?: 'backward' | 'forward' | 'none'
+  ): void {
+    super.setSelectionRange(start, end, direction);
+    this.selection = { start, end };
+  }
+
+  /** Checks for validity of the control and shows the browser message if it's invalid. */
+  public reportValidity() {
+    const state = this._value
+      ? this.parser.isValidString(this.input.value)
+      : this.input.reportValidity();
+    this.invalid = !state;
+    return state;
+  }
+
+  /**
+   * Sets a custom validation message for the control.
+   * As long as `message` is not empty, the control is considered invalid.
+   */
+  public setCustomValidity(message: string) {
+    this.input.setCustomValidity(message);
+    this.invalid = !this.input.checkValidity();
+  }
+
+  /** Check for validity of the control */
+  public checkValidity() {
+    if (this.disabled) {
+      return this.input.checkValidity();
+    }
+
+    if (!this._value && this.required) {
+      return false;
+    }
+
+    return (
+      this.input.checkValidity() && this.parser.isValidString(this.input.value)
+    );
```

Comment:
Close but no cigar 😄
An empty input (`!this._value === true`) that is not required will still hit the parser validation and become invalid, when it shouldn't. Can be adjusted this way:
```suggestion
    if (!this._value) {
      return !this.required;
    }

    return (
      this.input.checkValidity() && this.parser.isValidString(this.input.value)
    );
```
This can be adjusted with a focus check to go through the native check if you'd like.

Honestly, no idea how much we should care about calling the native check validity when the only thing it can do is required and only when the mask is not focused. We don't explicitly support the invalid event it produces and before we implement ElementInternals it can't be made consistent anyway.